spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binder/AbstractMessageChannelBinder.java

```diff
@@ -276,6 +269,31 @@ else if (e instanceof ProvisioningException) {
 		}
 	}
 
+	@Deprecated
+	private void enhanceMessageChannel(MessageChannel inputChannel) {
+		((AbstractMessageChannel)inputChannel).addInterceptor(new ChannelInterceptorAdapter() {
+			@SuppressWarnings("unchecked")
+			@Override
+			public Message<?> preSend(Message<?> message, MessageChannel channel) {
+				if (message.getPayload() instanceof byte[]) {
+					if (!message.getHeaders().containsKey(BinderHeaders.NATIVE_HEADERS_PRESENT)
+							&& EmbeddedHeaderUtils.mayHaveEmbeddedHeaders((byte[]) message.getPayload())) {
+						MessageValues messageValues;
```

Comment:
I am not sure why you deprecated `HeaderMode` (comment elsewhere in this review) but I think we only want to fall back to looking for embedded headers for binders that might support it. For example, the rabbit binder doesn't set the `NATIVE_HEADERS_PRESENT` - that is intended for binders that support embedded headers (e.g. Kafka for backwards compatibility).

Reply:
Should I implement now something myself in the Kinesis Binder to embed headers?
Thanks

Reply:
@artembilan it should just happen automatically if the headerMode is set.

This comment is redundant - discussed with @olegz - this method is only called if the header mode is `embeddedHeaders` - I was thrown off by the deprecation.

Reply:
But, @artembilan you might want to copy the embedded header test(s) from the kafka binder.

Reply:
Sure, @garyrussell ! Thank you for the pointer!

Although it's still not clear for me why we deprecate `HeaderMode` and this method still supporting `embeddedHeaders` mode. I might just miss some other hook on the matter...
Will investigate eventually.

Reply:
Ah, OK. @olegz has just addressed my concerns.
LGTM then 😄